File: lang/py/src/avro/schemaNormalization.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
# 
# http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""
Contains the SchemaNormalization class"""


try:
  import json
except ImportError:
  import simplejson as json
  
import hashlib
from avro import schema

class NoSuchAlgorithmException(Exception):
    pass

class SchemaNormalization:
    def __init__(self):
        #Initialize EMPTY64 and FP_TABLE
        self.EMPTY64 = 0xc15d213aa4d7a795        
        self.FP_TABLE=[]
    
        for i  in xrange (0,256):
          fp = i;
          for j in xrange(0,8):
            mask = -(fp & 1L);
            fp = (fp >> 1) ^ (self.EMPTY64 & mask);
          self.FP_TABLE.append(fp)

        self.env={}


    def toParsingForm(self, schema):
        """
        Transform schema instance to its canonical json string .
          @arg schema: avro schema instance
          @return: the canonical form of the avro schema
        """
        self.env={}
        return self._build(schema.to_json());

    def fingerprint(self,fpName, data):
        """
        Process hash value of data using custom hash algorithm
          @arg fpName: hash algorithm : CRC-64-AVRO or MD5 or SHA-1 or SHA-224 or SHA-256 or SHA-384 or SHA-512
          @return: hash value in hexa
        """
        if (fpName.lower() == "crc-64-avro"):
          fp = self.fingerprint64(data);
          return fp
    
        m = None
        if(fpName.lower() == "md5") :        m = hashlib.md5()
        elif(fpName.lower() == "sha-1") :     m = hashlib.sha1()
        elif(fpName.lower() == "sha-224") :   m = hashlib.sha224()
        elif(fpName.lower() == "sha-256") :  m = hashlib.sha256()
        elif(fpName.lower() == "sha-384") :  m = hashlib.sha384()
        elif(fpName.lower() == "sha-512") :  m = hashlib.sha512()
        
        if m is not None:
            m.update(data)
            return m.hexdigest();
        else:
            raise NoSuchAlgorithmException("Unknown hash algorithm : '%s'"%(fpName))


    def parsingFingerprint(self,fpName, schema):
        """
        Compute hash value of a schema canonical string using custom hash algorithm
          @arg schema: avro schema instance
          @return: the hash value of the canonical schema string
        """
        return self.fingerprint(fpName, self.toParsingForm(schema))

    def parsingFingerprint64(self,schema):
        """
        Compute hash value of a schema canonical string using CRC-64-AVRO hash algorithm
          @arg schema: avro schema instance
          @return: the hash value of the canonical schema string
        """
        return self.fingerprint64(self.toParsingForm(schema))
    
    def fingerprint64(self,data):
        """
        Compute CRC-64-AVRO hash algorithm for a string
          @arg data: string
          @return: the hash value of the string
        """
        result = self.EMPTY64;
        for b in data:
          result = (result >> 8) ^ self.FP_TABLE[(result ^ ord(b)) & 0xff];
        return hex(result)

    def _build(self,jsonobj):
        """
        Create the canonical avro schema string (json) from a complete avro schema string
        @arg json: json string representing an avro shema
        @arg env: environment dict 
        @return: canonical avro schema string
        """
        

        #print jsonobj
        #print type(jsonobj)
             
        o=""
        firstTime = True;
        st=""

        if jsonobj is None:
          st='"null"'
        elif type(jsonobj)==unicode:
          st=jsonobj
        elif type(jsonobj)==dict:
          st = jsonobj[u'type']
          if type(st)==dict:
            return self._build(jsonobj[u'type'])
          elif type(st)==list:
            return self._build(jsonobj[u'type'])
        elif type(jsonobj)==list:
          st = u'union'
        

        if(st==u"union"):
          o+='['
          for b in jsonobj:
            if not firstTime:
              o+=','
            else :
              firstTime = False;
            o+=self._build(b)
          o+=']'
          return o

        elif st in [u"array",u"map"]:
          o+="{\"type\":\""+st+"\""
          if st == u"array":
            o+=",\"items\":"
            o+=self._build( jsonobj[u"items"])
          else:
            o+=",\"values\":"
            o+=self._build( jsonobj[u"values"])
          o+="}"
          return o
        elif st in [u"enum",u"fixed",u"record",u"error"]:
          name = jsonobj[u"name"]
          
          if u"namespace" in jsonobj and len(jsonobj[u"namespace"])>0:
            #schema has a namespace
            if (name.find('.') > 0):
              #name already have a qualified name with namespace
              pass
            else:
              name=jsonobj[u"namespace"]+"."+name# namespace + full name
                
                
            
          if name in self.env:
            o+=self.env[name]
            return o;
                     
          qname = "\""+name+"\"";
          self.env[name] = qname;
          o+="{\"name\":"+qname;
          o+=",\"type\":\""+st+"\"";
          if st == u"enum":
            o+=",\"symbols\":[";
            for enumSymbol in jsonobj[u"symbols"]:
              if not firstTime:
                o+=',';
              else :
                firstTime = False;
              o+='"'+enumSymbol+'"';
            o+="]";
          elif st == u"fixed":
            o+=",\"size\":"+str(jsonobj[u"size"]);
          else :# st == u"record":
            o+=",\"fields\":[";
            for  f in  jsonobj[u"fields"]:
              if not firstTime:
                  o+=',';
              else:
                  firstTime = False;
              o+="{\"name\":\""+f[u"name"]+"\"";
              o+=",\"type\":"
              o+=self._build(f)
              o+="}";
            o+="]";
        elif st in schema.PRIMITIVE_TYPES:
          # boolean, bytes, double, float, int, long, null, string
          o+='"%s"'%(st)
          return o
        elif st in self.env:
          o+='"%s"'%(st)
          return o
        o+="}"
        return o

File: lang/java/avro/src/main/java/org/apache/avro/SchemaNormalization.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.avro;

import org.apache.avro.util.internal.JacksonUtils;
import java.util.Arrays;
import java.util.LinkedHashSet;
import java.util.Map;
import java.util.HashMap;
import java.util.TreeSet;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

/**
 * Collection of static methods for generating the parser canonical form of
 * schemas (see {@link #toParsingForm}), standard canonical form of schemas (see
 * {@link #toCanonicalForm}) with user defined properties and fingerprints of
 * canonical forms ({@link #fingerprint}).
 */
public class SchemaNormalization {

  private SchemaNormalization() {
  }

  /**
   * Returns "Parsing Canonical Form" of a schema as defined by Avro spec.
   */
  public static String toParsingForm(Schema s) {
    return toNormalizedForm(s, true, new LinkedHashSet<>());
  }

  /**
   * Returns "Standard Canonical Form" of a schema as defined by Avro spec.
   */
  public static String toCanonicalForm(Schema s) {
    return toCanonicalForm(s, new LinkedHashSet<>());
  }

  /**
   * Returns "Standard Canonical Form" of a schema as defined by Avro spec with
   * additional user standard properties.
   */
  public static String toCanonicalForm(Schema s, LinkedHashSet<String> properties) {
    LinkedHashSet<String> reservedProperties = new LinkedHashSet<>(Arrays.asList("name", "type", "fields", "symbols",
        "items", "values", "logicalType", "size", "order", "doc", "aliases", "default"));
    properties.removeAll(reservedProperties);
    return toNormalizedForm(s, false, properties);
  }

  private static String toNormalizedForm(Schema s, Boolean ps, LinkedHashSet<String> aps) {
    try {
      Map<String, String> env = new HashMap<>();
      return build(env, s, new StringBuilder(), ps, aps).toString();
    } catch (IOException e) {
      // Shouldn't happen, b/c StringBuilder can't throw IOException
      throw new RuntimeException(e);
    }
  }

  /**
   * Returns a fingerprint of a string of bytes. This string is presumed to
   * contain a canonical form of a schema. The algorithm used to compute the
   * fingerprint is selected by the argument <i>fpName</i>. If <i>fpName</i>
   * equals the string <code>"CRC-64-AVRO"</code>, then the result of
   * {@link #fingerprint64} is returned in little-endian format. Otherwise,
   * <i>fpName</i> is used as an algorithm name for
   * {@link MessageDigest#getInstance(String)}, which will throw
   * <code>NoSuchAlgorithmException</code> if it doesn't recognize the name.
   * <p>
   * Recommended Avro practice dictates that <code>"CRC-64-AVRO"</code> is used
   * for 64-bit fingerprints, <code>"MD5"</code> is used for 128-bit fingerprints,
   * and <code>"SHA-256"</code> is used for 256-bit fingerprints.
   */
  public static byte[] fingerprint(String fpName, byte[] data) throws NoSuchAlgorithmException {
    if (fpName.equals("CRC-64-AVRO")) {
      long fp = fingerprint64(data);
      byte[] result = new byte[8];
      for (int i = 0; i < 8; i++) {
        result[i] = (byte) fp;
        fp >>= 8;
      }
      return result;
    }

    MessageDigest md = MessageDigest.getInstance(fpName);
    return md.digest(data);
  }

  /**
   * Returns the 64-bit Rabin Fingerprint (as recommended in the Avro spec) of a
   * byte string.
   */
  public static long fingerprint64(byte[] data) {
    long result = EMPTY64;
    for (byte b : data)
      result = (result >>> 8) ^ FP64.FP_TABLE[(int) (result ^ b) & 0xff];
    return result;
  }

  /**
   * Returns {@link #fingerprint} applied to the parsing canonical form of the
   * supplied schema.
   */
  public static byte[] parsingFingerprint(String fpName, Schema s) throws NoSuchAlgorithmException {
    return fingerprint(fpName, toParsingForm(s).getBytes(StandardCharsets.UTF_8));
  }

  /**
   * Returns {@link #fingerprint64} applied to the parsing canonical form of the
   * supplied schema.
   */
  public static long parsingFingerprint64(Schema s) {
    return fingerprint64(toParsingForm(s).getBytes(StandardCharsets.UTF_8));
  }

  private static Appendable build(Map<String, String> env, Schema s, Appendable o, Boolean ps,
      LinkedHashSet<String> aps) throws IOException {
    boolean firstTime = true;
    Schema.Type st = s.getType();
    LogicalType lt = null;
    if (!ps)
      lt = s.getLogicalType();
    switch (st) {
    default: // boolean, bytes, double, float, int, long, null, string
      if (!ps && lt != null)
        return writeLogicalType(s, lt, o, aps);
      else
        return o.append('"').append(st.getName()).append('"');

    case UNION:
      o.append('[');
      for (Schema b : s.getTypes()) {
        if (!firstTime)
          o.append(',');
        else
          firstTime = false;
        build(env, b, o, ps, aps);
      }
      return o.append(']');

    case ARRAY:
    case MAP:
      o.append("{\"type\":\"").append(st.getName()).append("\"");
      if (st == Schema.Type.ARRAY)
        build(env, s.getElementType(), o.append(",\"items\":"), ps, aps);
      else
        build(env, s.getValueType(), o.append(",\"values\":"), ps, aps);
      if (!ps)
        setSimpleProps(o, s.getObjectProps(), aps); // adding the reserved property if not parser canonical schema
      return o.append("}");

    case ENUM:
    case FIXED:
    case RECORD:
      String name = s.getFullName();
      if (env.get(name) != null)
        return o.append(env.get(name));
      String qname = "\"" + name + "\"";
      env.put(name, qname);
      o.append("{\"name\":").append(qname);
      o.append(",\"type\":\"").append(st.getName()).append("\"");
      if (st == Schema.Type.ENUM) {
        o.append(",\"symbols\":[");
        for (String enumSymbol : s.getEnumSymbols()) {
          if (!firstTime)
            o.append(',');
          else
            firstTime = false;
          o.append('"').append(enumSymbol).append('"');
        }
        o.append("]");
      } else if (st == Schema.Type.FIXED) {
        o.append(",\"size\":").append(Integer.toString(s.getFixedSize()));
        lt = s.getLogicalType();
        // adding the logical property
        if (!ps && lt != null)
          setLogicalProps(o, lt);
      } else { // st == Schema.Type.RECORD
        o.append(",\"fields\":[");
        for (Schema.Field f : s.getFields()) {
          if (!firstTime)
            o.append(',');
          else
            firstTime = false;
          o.append("{\"name\":\"").append(f.name()).append("\"");
          build(env, f.schema(), o.append(",\"type\":"), ps, aps);
          if (!ps)
            setFieldProps(o, f, aps); // if standard canonical form then add reserved properties
          o.append("}");
        }
        o.append("]");
      }
      if (!ps) {
        setComplexProps(o, s);
        setSimpleProps(o, s.getObjectProps(), aps);
      } // adding the reserved property if not parser canonical schema
      return o.append("}");
    }
  }

  private static Appendable writeLogicalType(Schema s, LogicalType lt, Appendable o, LinkedHashSet<String> aps)
      throws IOException {
    o.append("{\"type\":\"").append(s.getType().getName()).append("\"");
    // adding the logical property
    setLogicalProps(o, lt);
    // adding the reserved property
    setSimpleProps(o, s.getObjectProps(), aps);
    return o.append("}");
  }

  private static void setLogicalProps(Appendable o, LogicalType lt) throws IOException {
    o.append(",\"").append(LogicalType.LOGICAL_TYPE_PROP).append("\":\"").append(lt.getName()).append("\"");
    if (lt.getName().equals("decimal")) {
      LogicalTypes.Decimal dlt = (LogicalTypes.Decimal) lt;
      o.append(",\"precision\":").append(Integer.toString(dlt.getPrecision()));
      if (dlt.getScale() != 0)
        o.append(",\"scale\":").append(Integer.toString(dlt.getScale()));
    }
  }

  private static void setSimpleProps(Appendable o, Map<String, Object> schemaProps, LinkedHashSet<String> aps)
      throws IOException {
    for (String propKey : aps) {
      if (schemaProps.containsKey(propKey)) {
        String propValue = JacksonUtils.toJsonNode(schemaProps.get(propKey)).toString();
        o.append(",\"").append(propKey).append("\":").append(propValue);
      }
    }
  }

  private static void setComplexProps(Appendable o, Schema s) throws IOException {
    if (s.getDoc() != null && !s.getDoc().isEmpty())
      o.append(",\"doc\":\"").append(s.getDoc()).append("\"");
    if (s.getAliases() != null && !s.getAliases().isEmpty())
      o.append(",\"aliases\":").append(JacksonUtils.toJsonNode(new TreeSet<String>(s.getAliases())).toString());
    if (s.getType() == Schema.Type.ENUM && s.getEnumDefault() != null) {
      o.append(",\"default\":").append(JacksonUtils.toJsonNode(s.getEnumDefault()).toString());
    }
  }

  private static void setFieldProps(Appendable o, Schema.Field f, LinkedHashSet<String> aps) throws IOException {
    if (f.order() != null)
      o.append(",\"order\":\"").append(f.order().toString()).append("\"");
    if (f.doc() != null)
      o.append(",\"doc\":\"").append(f.doc()).append("\"");
    if (!f.aliases().isEmpty())
      o.append(",\"aliases\":").append(JacksonUtils.toJsonNode(new TreeSet<String>(f.aliases())).toString());
    if (f.defaultVal() != null)
      o.append(",\"default\":").append(JacksonUtils.toJsonNode(f.defaultVal()).toString());
    setSimpleProps(o, f.getObjectProps(), aps);
  }

  final static long EMPTY64 = 0xc15d213aa4d7a795L;

  /* An inner class ensures that FP_TABLE initialized only when needed. */
  private static class FP64 {
    private static final long[] FP_TABLE = new long[256];
    static {
      for (int i = 0; i < 256; i++) {
        long fp = i;
        for (int j = 0; j < 8; j++) {
          long mask = -(fp & 1L);
          fp = (fp >>> 1) ^ (EMPTY64 & mask);
        }
        FP_TABLE[i] = fp;
      }
    }
  }
}
